### ISSUE TITLE
Run the shellspec test framework

### ISSUE THREAD

**Issue by ko1nksm** [oilshell/oil#752]
Hi,

I am interested in run with [ShellSpec](https://github.com/shellspec/shellspec). 

ShellSpec is a testing framework for POSIX compliant shell script. One of the notable points is that most features are implemented by pure shell scripts instead of calling external commands.

It contains tests for ShellSpec itself. Therefore you can check if it works easily.

```sh
cd /shellspec/installed/directory
shellspec -s /path/to/osh
```

The currently code size is

- Total (includes comment and blank): about 7000
- NLOC (Non-comment Lines of Code): 5000 over
- LLOC (Logical Lines of Code):  4000 over

This is reported by [ShellMetrics](https://github.com/shellspec/shellmetrics) - Cyclomatic Complexity Analyzer. It is also implemented by shell script.

ShellSpec needs a requires correct implementation of errexit. (ref. #709)
(Note: Workarounds have been implemented for some shell flaws.)
So when the bosh ([Schily Bourne Shell](http://schilytools.sourceforge.net/bosh.html)) broke, [I could immediately notice it](https://sourceforge.net/p/schilytools/tickets/search/?q=schily&filter=%7B%22reported_by%22%3A%5B%22ko1nksm%22%5D%7D
).

Unfortunately, ShellSpec is for the POSIX shell, so it contains almost no bash syntax. However, you can check if osh is POSIX compliant.

I'm sorry, but I'm a little busy right now, so I may not be able to help enough. but I think ShellSpec can be used to test run under osh.

I ran ShellSpec on osh 0.8.pre4 and it showed as below.

```sh
  [ ${BASH_VERSION+x} ] &&:
  ^
[ eval at line 34 of /home/koichi/workspace/shellspec/lib/general.sh ]:34: fatal: Undefined variable
  [ ${ZSH_VERSION+x} ] &&:
  ^
[ eval at line 34 of /home/koichi/workspace/shellspec/lib/general.sh ]:34: fatal: Undefined variable
  [ ${YASH_VERSION+x} ] &&:
  ^
[ eval at line 34 of /home/koichi/workspace/shellspec/lib/general.sh ]:34: fatal: Undefined variable
  [ ${POSH_VERSION+x} ] &&:
  ^
[ eval at line 34 of /home/koichi/workspace/shellspec/lib/general.sh ]:34: fatal: Undefined variable
  [ ${KSH_VERSION+x} ] &&:
  ^
[ eval at line 34 of /home/koichi/workspace/shellspec/lib/general.sh ]:34: fatal: Undefined variable
      [ -u "$1" ] && chmod 0700 "$1"
      ^
/home/koichi/workspace/shellspec/lib/libexec/runner.sh:11: fatal: Id.BoolUnary_u isn't implemented
    ( umask 0077; mkdir "$1"
    ^
/home/koichi/workspace/shellspec/lib/libexec/runner.sh:8: fatal: Exiting with status 1 (subshell invoked from PID 32443)
```

Regards,

**Comment by andychu** [comment]
OK thanks for the info.  The first error is `-u` isn't implemented, and that's easy to fix.

I'm not sure why you are getting undefined var errors though.  Oil should be POSIX compatible there.

- errexit will be POSIX compatible!  The changes are hidden behind `shopt`.


It's good to know that it uses builtins rather than external commands... that could indeed be a good test case!

**Comment by ko1nksm** [comment]
Oh, sorry. I didn't report enough. I use `set -eu`.

```console
$ osh -c 'set -eu; echo "${FOO-foo}"; BAR=x; echo "${BAR+bar}"; echo "${BAZ+baz}"'
foo
bar
  set -eu; echo "${FOO-foo}"; BAR=x; echo "${BAR+bar}"; echo "${BAZ+baz}"
                                                        ^~~~
[ -c flag ]:1: fatal: Undefined variable
```

Should be

```console
$ dash -c 'set -eu; echo "${FOO-foo}"; BAR=x; echo "${BAR+bar}"; echo "${BAZ+baz}"'
foo
bar

```

An implementation omission? 🤔

**Comment by andychu** [comment]
Yes thank you, I have verified this.  I will fix it!

**Comment by andychu** [comment]
OK I've fixed these two issues in #753 and #754 

Testing on `master` should be easy with instructions here, should take 30 seconds on most Linux machines:

https://github.com/oilshell/oil/wiki/Contributing

Thanks for the report!  I should make another release in the next few weeks.

**Comment by ko1nksm** [comment]
Thank you for quick fix!

I have confirmed that these issues have been fixed. However ShellSpec still doesn't seem to work.

I'm not sure if this is due to errexit, but I will investigate and report again.

**Comment by andychu** [comment]
I just fixed a bug that affects the double quoted case `"${x+foo}"`:

https://github.com/oilshell/oil/commit/527aa5670f97ddafc4562514bfadb09bcf1fc5b2

Although the snippets above aren't double quoted so maybe it wasn't the cause ?

Anyway I'm about to make a release and the release process caught it.

**Comment by ko1nksm** [comment]
Unfortunately yet another error. I did some investigation.

I get an error message that should not be displayed.

```sh
$ ~/git/oil/bin/osh -c '(set -e; false); echo ok'
  (set -e; false); echo ok
           ^~~~~
[ -c flag ]:1: fatal: Exiting with status 1 (command in PID 27321)
ok
```

The function name `func` causes an error. This is my favorite function name. 🙃 

```sh
$ ~/git/oil/bin/osh -c 'func() { :; }'
  func() { :; }
       ^
[ -c flag ]:1: Syntax error in expression
```

ShellSpec started to work when these two problems were avoided. However, some tests failed and stopped halfway. Perhaps need more investigation.

**Comment by ko1nksm** [comment]
Does not match the pattern with parentheses.

```sh
$ ~/git/oil/bin/osh -c 'case "foo()" in *\(\)) echo "should be match"; esac'
```

```
$ dash -c 'case "foo()" in *\(\)) echo "should be match"; esac'
should be match
```

**Comment by andychu** [comment]
1. Ooof!  The case problm is because Oil uses GNU libc to processes extended globs, and `*()` is an extended glob pattern!  Let me think about that.
2. `func` -- That is sort of intended and documented here: https://www.oilshell.org/release/0.8.pre2/doc/known-differences.html#oil-has-more-keywords-which-shadow-builtins-functions-and-commands .  Although more people have hit stuff like this so maybe we need to hide it behind a flag, e.g. #735 
3. The error message is also intended.... let me think about the first 2 before getting to this.

Thanks for the reports!

**Comment by andychu** [comment]
Extended globs are turned on here:

https://github.com/oilshell/oil/blob/master/native/libc.c#L68

But I guess we need a flag to only use extended globs if it's parsed STATICALLY, not dynamically... hm.

Or maybe it needs to be turned off in case statements only?

**Comment by ko1nksm** [comment]
**1.** Let me correct a little. It has the same meaning, but it was written as follows in the source code.

```sh
~/git/oil/bin/osh -c 'case "foo()" in *"()") echo "should be match"; esac'
```

Is this really an extended glob? I don't think it's an extended glob because of the quotes or the `\`.

**2.** Oh, sorry. I didn't think `func` was a keyword. I fixed ShellSpec.

**3.** Is it to make you aware unexpected exit?

I can suppress the display by the following method,

```sh
(set -e; false) 2>/dev/null
```

but, the intentional output will also disappear. umm....

```sh
(set -e; echo "some warnings" >&2; false) 2>/dev/null
```

Need a flags?

By the way, thanks to you, I noticed that ShellSpec fails when I enable extglob in zsh.

**Comment by andychu** [comment]
@ko1nksm I released the fixes toward running ShellSpec:

http://www.oilshell.org/blog/2020/05/translation-progress.html

I'm interested in more bug reports along these lines.

Is the output message blocking further testing?  I think we could have a `shopt` option for it, but I may want to do something more general too.

i.e. not sure if I want a flag simply for turning off a single error message

It might relate to #737 for optional color, i.e. optional  verbosity too

**Comment by ko1nksm** [comment]
First, Fixing for glob related issues seems to perfect. Fixing for related to the `func` keyword also seems to be ok. Thank you! But I rename function name to make ShellSpec does not depend on shell options as much as possible.

As for the error message related below still a problem in several places.

```sh
$ ~/git/oil/bin/osh -c '(set -e; false)'
  (set -e; false)
           ^~~~~
[ -c flag ]:1: fatal: Exiting with status 1 (command in PID 27321)
```

To be exact, the ShellSpec assertions works, but it fails due to the unexpected output.

The ShellSpec is treat as an error the unexpected output. The unexpected output is output that is caused by a bug in the framework or extensions, incorrect DSL usage, etc. This is important as a last resort to detect unexpected problems and cannot be suppressed (by `2>/dev/null`) without carefully.

ShellSpec uses code with like the following structure in several places.

```sh
set +e
( set -e
  shellspec_call "$@" # $@ is command/function that user wants to execute.
)
if [ $? -ne 0 ]; then ...
```

If the user-specified function exit by error, an unexpected message (`fatal: Exiting with ...`) is output. I'm still not sure if I can handle the error messages from osh and from user-specified functions separately without compromising the reliability of the framework.


From here, I will explain my opinion. Please note that this is not the conclusion after a thorough investigation.

The much of the trouble with ShellSpec are expected to be messages when the subshell stopped, not when script stopped. In the following example, "fatal: Exiting with status 1" is output. However the script continues to run and "end" is also output.

```sh
$ ~/git/oil/bin/osh -c '(set -e; false); echo end'
  (set -e; false); echo end
           ^~~~~
[ -c flag ]:1: fatal: Exiting with status 1 (command in PID 14862)
end
```

I think this message is misleading, because may be using a subshell to not stop on any error and the script has not exiting. This message should only be output when the script is stopped.

I can expect many of my problems to be solved by output this message only when the script is stopped, but I still have concerns.

I agree that this message is useful. However it should treat as a debug message. As you know, other shells do not output anything when an error occurs. I regard that the output to stderr is one of the specifications of command and can be completely controlled by the developer. Therefore, by default, only the messages that the developer intentionally outputs should be output.

In other words, I think you need a (best practice) flag to output debug messages instead of suppressing specified messages. (I'm sorry if it already exists.) Of course, this applies to only osh for compatibility, not oil.

**Comment by andychu** [comment]
OK, we can suppress it for sure -- I will think about the right mechanism for it.  Thanks for the reports!

**Comment by andychu** [comment]
OK I just added `shopt -u verbose_errexit` to silence the message.  I'm curious if ShellSpec can now run.

I think the default should be verbose, but if it a lot of other people run into problems, we can revisit it

**Comment by andychu** [comment]
Thanks!

1. This is a bug and I added it to the spec tests, although I'm not sure the cause
2. Yes `set -a` should eventually be done since it's POSIX, but I haven't encountered a script that uses it
3. I `-f` and `-C`, although there are likely still more issues
4 and 5. These are because of https://www.oilshell.org/release/0.8.pre5/doc/known-differences.html#break-continue-return-are-keywords-not-builtins

Although 5 is maybe slightly different... hm.

The __SOURCE__ idiom is interesting and I haven't seen that... It might be possible to fill these in dynamically too, not sure.  Let me fix the bug 1 first.

**Comment by andychu** [comment]
Actually 5 is a different bug than 4... looking into it...

**Comment by andychu** [comment]
OK I fixed (5)  as #774.  (1) still needs to be fixed!

**Comment by andychu** [comment]
I fixed #775 , the errexit divergence.  Great report!  

I appreciate any more testing.  A lot has changed in the last few months.

https://www.oilshell.org/release/0.8.3/

**Comment by ko1nksm** [comment]
Hi, it's been a while.
Congratulations on the release of oil 0.8.7.

I have not been able to test osh for a while due to lack of time. During that time, ShellSpec has also changed a lot. osh version 0.8.6 gave the error `Error closing descriptor 5: Bad file descriptor`, but it has been fixed in 0.8.7, and ShellSpec's self tests now work without any problems! 🎉

Unfortunately, it doesn't work correctly when `strict_errexit` is enabled, but these are issues that need to be addressed on the ShellSpec side, and I will work on them in the future.

**Comment by andychu** [comment]
Great, glad to hear it!  Thanks for trying it.

I'm not sure why it started working, but there are some release notes here (requires login):

https://oilshell.zulipchat.com/#narrow/stream/121540-oil-discuss/topic/Release.200.2E8.2E7

And a new feature in 0.8.7 which I think shell experts should find useful:

http://www.oilshell.org/release/latest/doc/xtrace.html

You can turn `oil:basic` on, or just turn on `xtrace_rich` and turn off `xtrace_details`.  Let me know if this makes sense!  I'm looking for feedback based on real use cases.

**Comment by ko1nksm** [comment]
The error `Error closing descriptor 5: Bad file descriptor` was raised by a recent ShellSpec fix that strictly closes unnecessary file descriptors that were not closed before. I checked the problem with 0.8.6 and immediately retested with master and it did not occur.

`xtrace_rich` and `xtrace_details` is a very interesting feature. ShellSpec has a debugging feature, the `-x` (`-X`) option, which traces only the function to be tested. For now it is still just turning on `xtrace`, but there are plans to add more information to make it easier to debug. I plan to try out those features in the near future.

ShellSpec captures the stderr so that it can be tested. However, when `xtrace` is turned on, the stderr cannot  be tested because the trace information will be mixed. I don't think osh supports the `SHX_descriptor` (`BASH_XTRACEFD`) feature yet (let me know if I'm wrong), but when it is implemented, it will be possible to test and trace at the same time. As a side note, `BASH_XTRACEFD` is also supported by recent BusyBox `ash`. This means that currently only bash and ash can run tests and traces at the same time.

**Comment by andychu** [comment]
Ah I didn't know that ash supports BASH_XTRACEFD.  That's a good argument for Oil to support it.

I mentioned that and other ideas on #891.  Feel free to add more feedback there!

I also think that `SHX_indent` and `SHX_pid_str` should be automatically exported, so that shell scripts that invoke other shell scripts can have unified traces!  I think it will work.

**Comment by ko1nksm** [comment]
One more side note, There is a request for zsh to implement `ZSH_XTRACEFD` as well. However, it seems that not much progress has been made.

- https://www.zsh.org/mla/workers/2019/msg00353.html
- https://www.zsh.org/mla/workers/2020/msg00533.html

(ShellSpec also secretly supports `ZSH_XTRACEFD` :P)
